tests/unit/testing/test_ambient_integration.py

```diff
@@ -0,0 +1,641 @@
+        raise_on_blocked=False,
+        raise_on_error=True,
+        timeout=900,
+    )
+
+
+@pytest.mark.asyncio
+async def test_deploy_and_integrate_service_mesh_charms_default_channel():
+    """Test deploy service mesh charms with default channel."""
+    app_name = "my-app"
+    app = Mock(spec_set=Application)()
+    app.name = app_name
+
+    model = AsyncMock(spec_set=Model)
+    model.applications = {app_name: app}
+    model.name = "test-model"
+
+    await deploy_and_integrate_service_mesh_charms(app_name, model)
+
+    # Verify default channel is used
+    model.deploy.assert_any_await(ISTIO_K8S_APP, channel="2/edge", trust=True)
+
+
+@pytest.mark.asyncio
+async def test_deploy_and_integrate_service_mesh_charms_no_ingress():
+    """Test deploy service mesh charms without ingress integration."""
+    app_name = "my-app"
+    app = Mock(spec_set=Application)()
+    app.name = app_name
+
+    model = AsyncMock(spec_set=Model)
+    model.applications = {app_name: app}
+    model.name = "test-model"
+
+    await deploy_and_integrate_service_mesh_charms(app_name, model, relate_to_ingress=False)
+
+    # Verify all three Istio charms were deployed
+    assert model.deploy.call_count == 3
+
+    # Verify only beacon integration (no ingress)
+    assert model.integrate.call_count == 1
+    model.integrate.assert_any_await(
+        f"{ISTIO_BEACON_K8S_APP}:service-mesh", f"{app_name}:service-mesh"
+    )
+
+
+@pytest.mark.asyncio
+async def test_deploy_and_integrate_service_mesh_charms_no_beacon():
+    """Test deploy service mesh charms without beacon integration."""
+    app_name = "my-app"
+    app = Mock(spec_set=Application)()
+    app.name = app_name
+
+    model = AsyncMock(spec_set=Model)
+    model.applications = {app_name: app}
+    model.name = "test-model"
+
+    await deploy_and_integrate_service_mesh_charms(app_name, model, relate_to_beacon=False)
+
+    # Verify all three Istio charms were deployed
+    assert model.deploy.call_count == 3
+
+    # Verify only ingress integration (no beacon)
+    assert model.integrate.call_count == 1
+    model.integrate.assert_any_await(
+        f"{ISTIO_INGRESS_K8S_APP}:istio-ingress-route", f"{app_name}:istio-ingress-route"
+    )
+
+
+@pytest.mark.asyncio
+async def test_deploy_and_integrate_service_mesh_charms_no_integrations():
+    """Test deploy service mesh charms without any integrations."""
+    app_name = "my-app"
+    app = Mock(spec_set=Application)()
+    app.name = app_name
+
+    model = AsyncMock(spec_set=Model)
+    model.applications = {app_name: app}
+    model.name = "test-model"
+
+    await deploy_and_integrate_service_mesh_charms(
+        app_name, model, relate_to_ingress=False, relate_to_beacon=False
+    )
+
+    # Verify all three Istio charms were deployed
+    assert model.deploy.call_count == 3
+
+    # Verify no integrations
+    model.integrate.assert_not_called()
+
+    # Still waits for idle
+    model.wait_for_idle.assert_awaited_once()
+
```

Comment:
suggestion (non-blocking): all the tests up to this points are so similar that I am thinking about parametrizing them into a unique one that calls `deploy_and_integrate_service_mesh_charm` with different `*args` and `**kwargs` defining the different test cases, so that we are still free to run the tested function for all of them with the same instruction